File: gto-support-util/src/main/java/org/ccci/gto/android/common/util/os/BundleUtils.java
package org.ccci.gto.android.common.util.os;

import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.TextUtils;

import org.ccci.gto.android.common.compat.util.LocaleCompat;
import org.jetbrains.annotations.Contract;

import java.lang.reflect.Array;
import java.util.Locale;

public class BundleUtils {
    @Nullable
    @SuppressWarnings({"unchecked", "SuspiciousSystemArraycopy"})
    public static <T extends Parcelable> T[] getParcelableArray(@NonNull final Bundle bundle,
                                                                @Nullable final String key,
                                                                @NonNull final Class<T> clazz) {
        final Parcelable[] raw = bundle.getParcelableArray(key);
        if (raw == null) {
            return null;
        }

        // copy all objects to typed array
        final T[] arr = (T[]) Array.newInstance(clazz, raw.length);
        System.arraycopy(raw, 0, arr, 0, raw.length);
        return arr;
    }

    public static void putEnum(@NonNull final Bundle bundle, @Nullable final String key,
                               @Nullable final Enum<?> value) {
        bundle.putString(key, value != null ? value.name() : null);
    }

    @Nullable
    public static <T extends Enum<T>> T getEnum(@NonNull final Bundle bundle, @NonNull final Class<T> type,
                                                @Nullable final String key) {
        return getEnum(bundle, type, key, null);
    }

    @Nullable
    @Contract("_, _, _, !null -> !null")
    public static <T extends Enum<T>> T getEnum(@NonNull final Bundle bundle, @NonNull final Class<T> type,
                                                @Nullable final String key, @Nullable final T defValue) {
        final String raw = bundle.getString(key);
        if (raw == null) {
            return defValue;
        }

        try {
            return Enum.valueOf(type, raw);
        } catch (final IllegalArgumentException e) {
            return defValue;
        }
    }

    public static void putLocale(@NonNull final Bundle bundle, @Nullable final String key,
                                 @Nullable final Locale locale) {
        bundle.putString(key, locale != null ? LocaleCompat.toLanguageTag(locale) : null);
    }

    @Nullable
    public static Locale getLocale(@NonNull final Bundle bundle, @Nullable final String key) {
        return getLocale(bundle, key, null);
    }

    @Nullable
    @Contract("_, _, !null -> !null")
    public static Locale getLocale(@NonNull final Bundle bundle, @Nullable final String key,
                                   @Nullable final Locale defValue) {
        final String raw = bundle.getString(key);
        if (raw == null) {
            return defValue;
        }
        return LocaleCompat.forLanguageTag(raw);
    }

    public static void putLocaleArray(@NonNull final Bundle bundle, @Nullable final String key,
                                      @Nullable final Locale[] locales) {
        putLocaleArray(bundle, key, locales, false);
    }

    /**
     * Store an array of Locales in the provided Bundle
     *
     * @param bundle       The bundle to store the locale array in
     * @param key          The key to store the locale array under
     * @param locales      The locales being put in the bundle
     * @param singleString Flag indicating if the locale array should be stored as a single string
     */
    public static void putLocaleArray(@NonNull final Bundle bundle, @Nullable final String key,
                                      @Nullable final Locale[] locales, final boolean singleString) {
        final String[] array;
        if (locales != null) {
            array = new String[locales.length];
            for (int i = 0; i < locales.length; i++) {
                array[i] = locales[i] != null ? LocaleCompat.toLanguageTag(locales[i]) : null;
            }
        } else {
            array = null;
        }

        if (singleString) {
            bundle.putString(key, array != null ? TextUtils.join(",", array) : null);
        } else {
            bundle.putStringArray(key, array);
        }
    }

    @Nullable
    public static Locale[] getLocaleArray(@NonNull final Bundle bundle, @Nullable final String key) {
        String[] raw = bundle.getStringArray(key);
        if (raw == null) {
            final String flat = bundle.getString(key);
            raw = flat != null ? TextUtils.split(flat, ",") : null;
        }

        final Locale[] locales;
        if (raw != null) {
            locales = new Locale[raw.length];
            for (int i = 0; i < raw.length; i++) {
                locales[i] = raw[i] != null ? LocaleCompat.forLanguageTag(raw[i]) : null;
            }
        } else {
            locales = null;
        }
        return locales;
    }
}

File: gto-support-util/src/main/java/org/ccci/gto/android/common/util/os/BundleUtils.kt
@file:JvmName("BundleUtils")

package org.ccci.gto.android.common.util.os

import android.os.Bundle
import java.util.Locale
import org.jetbrains.annotations.Contract

// region Enums

fun Bundle.putEnum(key: String?, value: Enum<*>?) = putString(key, value?.name)

@JvmOverloads
@Contract("_, _, _, !null -> !null")
fun <T : Enum<T>> Bundle.getEnum(type: Class<T>, key: String?, defValue: T? = null): T? {
    return try {
        getString(key)?.let { java.lang.Enum.valueOf<T>(type, it) } ?: defValue
    } catch (e: IllegalArgumentException) {
        defValue
    }
}

@Contract("_, _, !null -> !null")
inline fun <reified T : Enum<T>> Bundle.getEnum(key: String?, defValue: T? = null) =
    getEnum(T::class.java, key, defValue)

// endregion Enums

// region Locales
@Deprecated(
    "Since v3.11.2, use BundleKt.putLocale() instead",
    ReplaceWith("putLocale(key, locale)"),
    DeprecationLevel.ERROR
)
@JvmName("putLocale")
@Suppress("FunctionName")
internal fun Bundle.`-putLocale`(key: String?, locale: Locale?) = putLocale(key, locale)

@Deprecated(
    "Since v3.11.2, use BundleKt.getLocale() instead",
    ReplaceWith("getLocale(key, defValue)"),
    DeprecationLevel.ERROR
)
@JvmOverloads
@JvmName("getLocale")
@Suppress("FunctionName")
fun Bundle.`-getLocale`(key: String?, defValue: Locale? = null) = getLocale(key, defValue)

@Deprecated(
    "Since v3.11.2, use BundleKt.putLocaleArray() instead",
    ReplaceWith("putLocaleArray(key, locales, singleString)"),
    DeprecationLevel.ERROR
)
@JvmOverloads
@JvmName("putLocaleArray")
@Suppress("FunctionName")
fun Bundle.`-putLocaleArray`(key: String?, locales: Array<Locale>?, singleString: Boolean = false) =
    putLocaleArray(key, locales, singleString)

@Deprecated(
    "Since v3.11.2, use BundleKt.getLocaleArray() instead",
    ReplaceWith("getLocaleArray(key)"),
    DeprecationLevel.ERROR
)
@JvmName("getLocaleArray")
@Suppress("FunctionName")
fun Bundle.`-getLocaleArray`(key: String?) = getLocaleArray(key)
// endregion Locales
